modules/core/arrow-annotations-processor/src/main/java/arrow/renzu/RenzuGenerator.kt

```diff
@@ -0,0 +1,181 @@
+  annotatedList: List<AnnotatedInstance>,
+  val isolateForTests: Boolean = false) {
+
+  private val typeclassTree: MutableMap<TypeClass, Tuple2<Instances, Set<ParentTypeClass>>> =
+    normalizeTypeclassTree(annotatedList.map { Instance(it) })
+
+  private fun normalizeTypeclassTree(instances: List<Instance>)
+    : MutableMap<TypeClass, Tuple2<Instances, Set<ParentTypeClass>>> =
+    instances.fold(mutableMapOf()) { acc, instance ->
+      instance.implementedTypeclasses().forEach { tc ->
+        val typeclass = TypeClass(processor, tc.simpleName, tc)
+        val parentTypeClasses = typeclass.parentTypeClasses(tc)
+
+        acc.computeIfPresent(typeclass,
+          { _, value ->
+            Tuple2(
+              value._1 + setOf(instance),
+              value._2 + parentTypeClasses)
+          })
+        acc.putIfAbsent(typeclass,
+          Tuple2(setOf(instance), parentTypeClasses.toSet()))
+      }
+      acc
+    }
+
+  fun generate() {
+    val generatedDir = if (isolateForTests)
+      File("./infographic")
+    else
+      File("${recurseFilesUpwards("arrow").absolutePath}/infographic", "")
+        .also { it.mkdirs() }
+
+    val file = File(generatedDir, "arrow-infographic.txt")
+    if (!file.exists()) {
+      val globalStyles =
+        listOf("#font: Menlo") +
+          listOf("#fontSize: 10") +
+          listOf("#arrowSize: 1") +
+          listOf("#bendSize: 0.3") +
+          listOf("#lineWidth: 2") +
+          listOf("#padding: 8") +
+          listOf("#zoom: 1") +
+          listOf("#fill: #64B5F6")
+
+      file.appendText(globalStyles.joinToString(separator = "\n"))
+    }
+
+    val elementsToGenerate: List<String> = genDiagramRelations(typeclassTree)
+    val currentFileLines = file.readLines()
+
+    val source: String = elementsToGenerate
+      .toSet()
+      .filterNot { currentFileLines.contains(it) }
+      .joinToString(prefix = "\n", separator = "\n")
+
+    if (source != "\n") {
+      file.appendText(source)
+    }
+
+    processor.log("arrow-infographic generated: " + file.path)
+  }
+
+  /**
+   * Returns the UML text for the diagram to be rendered using nomnoml.
+   *
+   * Sample format for the output:
+   *
+   * #font: Menlo
+   * #fontSize: 10
+   * #arrowSize: 1
+   * #bendSize: 0.3
+   * #lineWidth: 2
+   * #padding: 8
+   * #zoom: 1
+   * #fill: #64B5F6
+   * #.typeclass: fill=#64B5F6 visual=database bold
+   * #.instances: fill=#B9F6CA visual=class italic bold dashed
+
+   * [<typeclass>Functor]<-[<typeclass>Applicative]
+   * [<typeclass>Applicative]<-[<typeclass>Monad]
+   * [<typeclass>Monad]<-[<instances>Monad Instances|NonEmptyList|Option|OptionT|SequenceK|State|StateT|Try|Either|EitherT|Eval|Id]
+   * [<typeclass>Applicative]<-[Something 2]
+   * [<typeclass>Applicative]<-[Something 3]
+   */
+  private fun genDiagramRelations(typeclassTree: MutableMap<TypeClass, Tuple2<Instances, Set<ParentTypeClass>>>)
+    : List<String> {
+    val relations = mutableListOf<String>()
```

Comment:
We can encode this entire method without mutability just flatMapping over `typeclassTree`